### ISSUE TITLE
datadogexporter is ignoring metricstransform 'update_label' action

### ISSUE THREAD

**Issue by emilgelman** [open-telemetry/opentelemetry-collector-contrib#6948]
**Describe the bug**
After upgrading the collector to version 0.41.0, metrics exported to Datadog have wrong tag values when using the `metricstransform` processor.
We are using `add_label` to add a default value, and `update_label` to allow overrides. The override is not taken into account.

For example, when reporting the following metric using the Go SDK:
```go
	c := metric.Must(meter).NewInt64UpDownCounter("test.counter")
	c.Add(ctx, 1, attribute.KeyValue{
		Key:   "override.service",
		Value: attribute.StringValue("othervalue"),
	})
```
The exported metric should have the tag `service:othervalue`, as seen in logging exporter:

```
Metric #0
Descriptor:
     -> Name: test.counter
     -> Description:
     -> Unit:
     -> DataType: Gauge
NumberDataPoints #0
Data point attributes:
     -> service: STRING(othervalue)
```

Instead, when looking at the Datadog app, the metric has the tag `service:test`

**What version did you use?**
v0.41.0

**What config did you use?**
```yaml
processors:
  metricstransform:
    transforms:
      include: .*
      match_type: regexp
      action: update
      operations:
        - action: 'add_label'
          new_label: "service"
          new_value: "test"
        - action: 'update_label'
          new_label: "service"
          label: "override.service"
```
**Environment**
Ubuntu 14

**Comment by jpkrohling** [comment]
cc @KSerrania @mx-psi @gbbr @knusbaum

**Comment by KSerrania** [comment]
Hi @emilgelman,

Thanks for reporting this issue!

I've tried reproducing your issue with the following setup:

<details>

<summary>main.go</summary>

```
package main

import (
	"context"
	"log"
	"time"

	"google.golang.org/grpc"

	"go.opentelemetry.io/otel/attribute"
	"go.opentelemetry.io/otel/exporters/otlp/otlpmetric/otlpmetricgrpc"
	"go.opentelemetry.io/otel/metric"
	"go.opentelemetry.io/otel/metric/global"
	"go.opentelemetry.io/otel/sdk/resource"
	semconv "go.opentelemetry.io/otel/semconv/v1.7.0"
	controller "go.opentelemetry.io/otel/sdk/metric/controller/basic"
	processor "go.opentelemetry.io/otel/sdk/metric/processor/basic"
	"go.opentelemetry.io/otel/sdk/metric/selector/simple"
)

// Initializes an OTLP exporter, and configures the corresponding trace and
// metric providers.
func initProvider() func() {
	ctx := context.Background()

	res, err := resource.New(ctx,
		resource.WithAttributes(
			semconv.DeploymentEnvironmentKey.String("dev"),
		),
	)
	  handleErr(err, "failed to create exporter")


	// If the OpenTelemetry Collector is running on a local cluster (minikube or
	// microk8s), it should be accessible through the NodePort service at the
	// `localhost:30080` endpoint. Otherwise, replace `localhost` with the
	// endpoint of your cluster. If you run the app inside k8s, then you can
	// probably connect directly to the service through dns
	exp, err := otlpmetricgrpc.New(
		ctx,
		otlpmetricgrpc.WithInsecure(),
		otlpmetricgrpc.WithEndpoint("otel-collector:30080"),
		otlpmetricgrpc.WithDialOption(grpc.WithBlock()),
	)
	handleErr(err, "failed to create exporter")

	cont := controller.New(
		processor.New(
			simple.NewWithHistogramDistribution(),
			exp,
		),
		controller.WithResource(res),
		controller.WithExporter(exp),
		controller.WithCollectPeriod(2*time.Second),
	)

	global.SetMeterProvider(cont.MeterProvider())
	handleErr(cont.Start(context.Background()), "failed to start controller")

	return func() {
		// Push any last metric events to the exporter.
		handleErr(cont.Stop(context.Background()), "failed to stop controller")
	}
}

func main() {
	log.Printf("Waiting for connection...")

	shutdown := initProvider()
	defer shutdown()

	meter := global.Meter("test-meter")
	counter := metric.Must(meter).NewInt64UpDownCounter("test.counter")

	for i := 0; i < 100; i++ {
		log.Printf("Sending test.counter (%d / 100)\n", i+1)
		counter.Add(context.Background(), 1, attribute.KeyValue{
			Key:   "override.service",
			Value: attribute.StringValue("othervalue"),
		})
		<-time.After(time.Second)
	}

	log.Printf("Done!")
}

func handleErr(err error, message string) {
	if err != nil {
		log.Fatalf("%s: %v", message, err)
	}
}
```
</details>

<details>

<summary>go.mod</summary>

```
module github.com/DataDog/otel-collector-e2e-tests/services/otlp-sender

go 1.17

require (
	go.opentelemetry.io/otel v1.2.0
	go.opentelemetry.io/otel/exporters/otlp/otlpmetric/otlpmetricgrpc v0.23.0
	go.opentelemetry.io/otel/metric v0.23.0
	go.opentelemetry.io/otel/sdk v1.0.0
	go.opentelemetry.io/otel/sdk/metric v0.23.0
	google.golang.org/grpc v1.40.0
)

require (
	github.com/cenkalti/backoff/v4 v4.1.1 // indirect
	github.com/golang/protobuf v1.5.2 // indirect
	github.com/grpc-ecosystem/grpc-gateway v1.16.0 // indirect
	go.opentelemetry.io/otel/exporters/otlp/otlpmetric v0.23.0 // indirect
	go.opentelemetry.io/otel/internal/metric v0.23.0 // indirect
	go.opentelemetry.io/otel/sdk/export/metric v0.23.0 // indirect
	go.opentelemetry.io/otel/trace v1.2.0 // indirect
	go.opentelemetry.io/proto/otlp v0.9.0 // indirect
	golang.org/x/net v0.0.0-20210917221730-978cfadd31cf // indirect
	golang.org/x/sys v0.0.0-20210921065528-437939a70204 // indirect
	golang.org/x/text v0.3.7 // indirect
	google.golang.org/genproto v0.0.0-20210921142501-181ce0d877f6 // indirect
	google.golang.org/protobuf v1.27.1 // indirect
)
```

</details>

<details>

<summary>OTel collector config</summary>

```
receivers:
  otlp:
    protocols:
      grpc:
        endpoint: "0.0.0.0:30080"

processors:
  metricstransform:
    transforms:
      include: .*
      match_type: regexp
      action: update
      operations:
        - action: 'add_label'
          new_label: "service"
          new_value: "test"
        - action: 'update_label'
          new_label: "service"
          label: "override.service"

exporters:
  file/metrics:
    path: /etc/otel/metrics.dat
    
  datadog:
    api:
      key: <api key>
    metrics:

service:
  telemetry:
    logs:
      level: info

  pipelines:
    metrics:
      receivers: [otlp]
      processors: [metricstransform]
      exporters: [datadog, file/metrics]
```

</details>

but I haven't been able to, the metrics I get have the correct tag:

![image](https://user-images.githubusercontent.com/26872252/147481492-5490235b-e0ea-4aa1-a017-5f50ae5b8bf6.png)
![image](https://user-images.githubusercontent.com/26872252/147481556-5bfed384-48d4-4256-89e0-53c813674b6d.png)

To help us investigate, could you give us the full test application & collector configuration used (with sensitive data such as API keys removed)?

Moreover, could you tell us whether you're able to reproduce this issue with the previous collector version you were using?

**Comment by emilgelman** [comment]
Hi @KSerrania 
Thank you for checking this.

Below is a complete example:
<details>
  <summary>main.go</summary>
  
  ```go
    package main

import (
	"context"
	"go.opentelemetry.io/otel/attribute"
	"go.opentelemetry.io/otel/sdk/metric/aggregator/histogram"
	"log"
	"time"

	"go.opentelemetry.io/otel"
	"go.opentelemetry.io/otel/exporters/otlp/otlpmetric"
	"go.opentelemetry.io/otel/exporters/otlp/otlpmetric/otlpmetrichttp"
	"go.opentelemetry.io/otel/metric"
	"go.opentelemetry.io/otel/metric/global"
	controller "go.opentelemetry.io/otel/sdk/metric/controller/basic"
	processor "go.opentelemetry.io/otel/sdk/metric/processor/basic"
	"go.opentelemetry.io/otel/sdk/metric/selector/simple"
)

func main() {
	ctx := context.Background()
	client := otlpmetrichttp.NewClient(otlpmetrichttp.WithInsecure(), otlpmetrichttp.WithEndpoint("0.0.0.0:55681"))
	exp, err := otlpmetric.New(ctx, client)
	if err != nil {
		log.Fatalf("Failed to create the collector exporter: %v", err)
	}
	defer func() {
		ctx, cancel := context.WithTimeout(ctx, time.Second)
		defer cancel()
		if err := exp.Shutdown(ctx); err != nil {
			otel.Handle(err)
		}
	}()

	pusher := controller.New(
		processor.NewFactory(
			simple.NewWithHistogramDistribution(histogram.WithExplicitBoundaries([]float64{2.0, 4.0, 8.0})),
			exp,
		),
		controller.WithExporter(exp),
		controller.WithCollectPeriod(time.Millisecond*100),
	)
	global.SetMeterProvider(pusher)

	if err := pusher.Start(ctx); err != nil {
		log.Fatalf("could not start metric controoler: %v", err)
	}

	defer func() {
		ctx, cancel := context.WithTimeout(ctx, time.Second)
		defer cancel()
		// pushes any last exports to the receiver
		if err := pusher.Stop(ctx); err != nil {
			otel.Handle(err)
		}
	}()

	meter := global.Meter("test-meter")
	c := metric.Must(meter).NewInt64UpDownCounter("test.counter.basic")
	c.Add(ctx, 5, attribute.KeyValue{
		Key:   "override.service",
		Value: attribute.StringValue("othervalue"),
	})
	time.Sleep(time.Second)
	client.Stop(ctx)
}
  ```
</details>

<details>
  <summary>go.mod</summary>
  
  ```go
    module test

go 1.15

require (
	github.com/davecgh/go-spew v1.1.1 // indirect
	github.com/kr/text v0.2.0 // indirect
	github.com/niemeyer/pretty v0.0.0-20200227124842-a10e7caefd8e // indirect
	go.opentelemetry.io/otel v1.3.0
	go.opentelemetry.io/otel/exporters/otlp/otlpmetric v0.26.0
	go.opentelemetry.io/otel/exporters/otlp/otlpmetric/otlpmetrichttp v0.26.0
	go.opentelemetry.io/otel/metric v0.26.0
	go.opentelemetry.io/otel/sdk/metric v0.26.0
	golang.org/x/sys v0.0.0-20210616094352-59db8d763f22 // indirect
	golang.org/x/text v0.3.2 // indirect
	gopkg.in/check.v1 v1.0.0-20200227125254-8fa46927fb4f // indirect
	gopkg.in/yaml.v3 v3.0.0-20210107192922-496545a6307b // indirect
)
  ```
</details>
<details>
  <summary>OTel collector config</summary>
  
  ```yaml
receivers:
  otlp:
    protocols:
      http:
        endpoint: 0.0.0.0:55681
processors:
  metricstransform:
    transforms:
      include: .*
      match_type: regexp
      action: update
      operations:
        - action: 'add_label'
          new_label: "service"
          new_value: "test"
        - action: 'update_label'
          new_label: "service"
          label: "override.service"

exporters:
  datadog:
    hostname: "####"
    env: "prod"
    use_resource_metadata: false
    send_metadata: false
    api:
      key: "####"
      site: datadoghq.eu
    metrics:
      send_monotonic_counter: false
      resource_attributes_as_tags: true
      instrumentation_library_metadata_as_tags: true
      histograms:
        send_count_sum_metrics: true
    retry_on_failure:
      enabled: true
    sending_queue:
      enabled: false
  logging:
    loglevel: debug


service:
  pipelines:
    metrics:
      receivers: [otlp]
      processors: [metricstransform]
      exporters: [datadog, logging]

  ```
</details>

It looks like in version `0.41.0` I'm getting `override.service` as a standalone tag.

In addition, running this with collector version `0.40.0` does not reproduce the issue.

**Comment by KSerrania** [comment]
Hey @emilgelman,

Thanks for the details! To get more data, would you mind setting the collector log level to debug (by adding this to the collector configuration:
```
service:
  telemetry:
    logs:
      level: debug
  pipelines:
    <previous pipelines configuration>
```

in both the `v0.40.0` and `v0.41.0` test applications you made, and send us the full collector output for both (including what the logging exporter outputs, and without sensitive data)?

Thanks!

**Comment by emilgelman** [comment]
Hi @KSerrania, sorry for the delay.
After running additional tests, I was able to narrow down the problem. The previous configuration I've provided is missing some details (I apologize for that, got confused myself), so I'll try to provide as much information as I can now.

Our issue started when we upgraded from collector version `0.36.0` to `0.41.0`. We skipped the versions in between because of scale issues, which are now fixed in `0.41.0`. 
This issue can be reproduced starting collector version `0.40.0`.

The services that were affected once we upgraded are services which send metrics using `statsd`.
In addition, this only happens when we define the attributes on the SDK `Controller` level, and not on the `Instrument` itself. See the code example below where we define the `override.service` attribute.

Our configuration in `0.36.0` is as follows:
<details>
  <summary>0.36.0 collector</summary>
  
  ```yaml
   # Open Telemetry Collector Version: 0.36.0
receivers:
  otlp:
    protocols:
      http:
        endpoint: 0.0.0.0:55681

  statsd:
    endpoint: 0.0.0.0:8127
    aggregation_interval: 60s

processors:
  metricstransform:
    transforms:
      include: .*
      match_type: regexp
      action: update
      operations:
        - action: 'add_label'
          new_label: service
          new_value: test

        - action: 'update_label'
          new_label: service
          label: override.service
  resource:
    attributes:
      - key: service
        value: test
        action: 'upsert'

      - key: service
        from_attribute: override.service
        action: 'insert'

exporters:
  datadog:
    hostname: "####"
    env: "prod"
    use_resource_metadata: false
    send_metadata: false
    api:
      key: "####"
      site: datadoghq.eu
    metrics:
  #      send_monotonic_counter: false
  #      resource_attributes_as_tags: true
  #      instrumentation_library_metadata_as_tags: true
  logging:
    loglevel: debug

service:
  pipelines:
    metrics:
      receivers: [ otlp,statsd ]
      processors: [ metricstransform, resource ]
      exporters: [ datadog,logging ]

  ```
</details>

Notice that we are using both `metricstransform` and `resource` processor.

Once we upgrade to `0.41.0`, we have enabled `exporters.datadog.metrics.resource_attributes_as_tags=true`, to overcome the issue of missing attribute tags. Full configuration below:

<details>
  <summary>0.41.0 config</summary>
  
  ```yaml
    # Open Telemetry Collector Version: 0.41.0
receivers:
  otlp:
    protocols:
      http:
        endpoint: 0.0.0.0:55681

  statsd:
    endpoint: 0.0.0.0:8127
    aggregation_interval: 60s

processors:
  metricstransform:
    transforms:
      include: .*
      match_type: regexp
      action: update
      operations:
        - action: 'add_label'
          new_label: service
          new_value: test

        - action: 'update_label'
          new_label: service
          label: override.service
  resource:
    attributes:
      - key: service
        value: test
        action: 'upsert'

      - key: service
        from_attribute: override.service
        action: 'insert'

exporters:
  datadog:
    hostname: "####"
    env: "prod"
    use_resource_metadata: false
    send_metadata: false
    api:
      key: "####"
      site: datadoghq.eu
    metrics:
        send_monotonic_counter: true
        resource_attributes_as_tags: true
        instrumentation_library_metadata_as_tags: true
  logging:
    loglevel: debug

service:
  pipelines:
    metrics:
      receivers: [ otlp,statsd ]
      processors: [ metricstransform, resource ]
      exporters: [ datadog,logging ]
  ```
</details>

After this change, clients that were using `statsd` failed overriding their service tag using the `override.service` attribute. Example Go application:

<details>
  <summary>main.go</summary>
  
  ```go
    package main

import (
	"context"
	"go.opentelemetry.io/otel"
	"go.opentelemetry.io/otel/metric/global"
	"go.opentelemetry.io/otel/sdk/metric/aggregator/histogram"
	"go.opentelemetry.io/otel/sdk/metric/selector/simple"
	"log"
	"time"

	"go.opentelemetry.io/contrib/exporters/metric/dogstatsd"
	"go.opentelemetry.io/otel/attribute"
	"go.opentelemetry.io/otel/metric"
	controller "go.opentelemetry.io/otel/sdk/metric/controller/basic"
	processor "go.opentelemetry.io/otel/sdk/metric/processor/basic"
	"go.opentelemetry.io/otel/sdk/resource"
)

func main() {
	ctx := context.Background()
	exp, err := dogstatsd.NewRawExporter(dogstatsd.Config{
		URL: "udp://127.0.0.1:8127",
	})
	if err != nil {
		panic(err)
	}

	pusher := controller.New(
		processor.NewFactory(
			simple.NewWithHistogramDistribution(histogram.WithExplicitBoundaries([]float64{2.0, 4.0, 8.0})),
			exp,
		),
		controller.WithExporter(exp),
		controller.WithCollectPeriod(time.Millisecond*100),
		controller.WithResource(resource.NewSchemaless(attribute.KeyValue{
			Key:   "override.service",
			Value: attribute.StringValue("othervalue"),
		})),
	)
	global.SetMeterProvider(pusher)

	if err := pusher.Start(ctx); err != nil {
		log.Fatalf("could not start metric controoler: %v", err)
	}

	defer func() {
		ctx, cancel := context.WithTimeout(ctx, time.Second)
		defer cancel()
		// pushes any last exports to the receiver
		if err := pusher.Stop(ctx); err != nil {
			otel.Handle(err)
		}
	}()

	meter := global.Meter("test-meter")
	c := metric.Must(meter).NewInt64UpDownCounter("test.31.statsd.otel40.3.on")
	c.Add(ctx, 5)
	time.Sleep(time.Second)
}
  ```
</details>

<details>
  <summary>go.mod</summary>
  
  ```go
    module main

go 1.16

require (
	go.opentelemetry.io/contrib/exporters/metric/dogstatsd v0.27.0
	go.opentelemetry.io/otel v1.2.0
	go.opentelemetry.io/otel/metric v0.25.0
	go.opentelemetry.io/otel/sdk v1.2.0
	go.opentelemetry.io/otel/sdk/metric v0.25.0
)
  ```
</details>

In addition, it looks like for OTLP(http) our override on resource attributes never worked, even in `0.36.0`. I am still looking into this, as this may be a different issue.

**Comment by KSerrania** [comment]
Hey @emilgelman,

Thanks for the additional context! Before investigating more, I'd like to confirm something with you. I reproduced the new setup you've sent with v0.41.0 of the collector, and the logging exporter is telling me that it receives the following metric:

```
2022-01-03T10:32:16.828Z	INFO	loggingexporter/logging_exporter.go:54	MetricsExporter	{"#metrics": 1}
2022-01-03T10:32:16.830Z	DEBUG	loggingexporter/logging_exporter.go:64	ResourceMetrics #0
Resource SchemaURL:
Resource labels:
     -> service: STRING(test)
InstrumentationLibraryMetrics #0
InstrumentationLibraryMetrics SchemaURL:
InstrumentationLibrary
Metric #0
Descriptor:
     -> Name: test.31.statsd.otel40.3.on
     -> Description:
     -> Unit:
     -> DataType: Gauge
NumberDataPoints #0
Data point attributes:
     -> service: STRING(othervalue)
StartTimestamp: 2022-01-03 10:31:16.828608344 +0000 UTC
Timestamp: 2022-01-03 10:31:18.800568537 +0000 UTC
Value: 5
```

Could you confirm that you are getting something similar on your side? In particular, are the `Resource labels` and `Data point attributes` the same?

**Comment by emilgelman** [comment]
Hi @KSerrania,

I can confirm it looks identical to your output:
```
2022-01-03T11:08:49.582Z        INFO    loggingexporter/logging_exporter.go:54  MetricsExporter {"#metrics": 56}
2022-01-03T11:08:49.583Z        DEBUG   loggingexporter/logging_exporter.go:64  ResourceMetrics #0
Resource SchemaURL:
Resource labels:
     -> service: STRING(test)
InstrumentationLibraryMetrics #0
InstrumentationLibraryMetrics SchemaURL:
InstrumentationLibrary
Metric #34
Descriptor:
     -> Name: test.statsd.jan3
     -> Description:
     -> Unit:
     -> DataType: Gauge
NumberDataPoints #0
Data point attributes:
     -> service: STRING(othervalue)
 StartTimestamp: 2022-01-03 11:07:49.483486138 +0000 UTC
Timestamp: 2022-01-03 11:08:48.061374699 +0000 UTC
Value: 5
```

Here is what we think happens: since `StatsD` does not support `Resource` level attributes, the [dogstatsd exporter](go.opentelemetry.io/contrib/exporters/metric/dogstatsd) exports resource attributes as data point attributes.

When these get to the Collector, our `metrictransformer` adds a tag of `service:override` to the datapoint.
When it reaches the `Datadog exporter`, because of `resource_attributes_as_tags=true`, the exporter gets the service tag value from the resource processor, which is `test` - since no resource override was provided.

In short, is it possible that when specifying`resource_attributes_as_tags=true`, resource attributes take precendence over data point attributes? If yes, can this be changed/configured otherwise? I think that the tag value on the data point should take precendence.

Does that make sense to you?

**Comment by KSerrania** [comment]
> In short, is it possible that when specifying resource_attributes_as_tags=true, resource attributes take precendence over data point attributes?

Yes, I believe that is the case, and this explains what's been happening in this issue.

> If yes, can this be changed/configured otherwise?

The `resource_attributes_as_tags` option is powered by the `resourcetotelemetry` OTel helper (see [here](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/2f11be4a8f16a8262470ea6d66ab14d406780ff6/exporter/datadogexporter/factory.go#L170-L171)).

The `resourcetotelemetry` helper adds resource labels as datapoint-level attributes with an `upsert` (see [here](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/2f11be4a8f16a8262470ea6d66ab14d406780ff6/pkg/resourcetotelemetry/resource_to_telemetry.go#L101-L106)), so resource labels take precedence over datapoint-level attributes.

The `resourcetotelemetry` documentation does not mention that this can be changed through any config option, therefore I don't think the behavior you want (resource labels are taken into account, _but_ do not take precedence over datapoint-level attributes) can be configured in the current version.

>  I think that the tag value on the data point should take precendence.

In that case, I think a new issue should be raised on the `resourcetotelemetry` helper to propose this alternative behavior (likely through a config option). Then, if accepted and implemented, we could expose that option in our exporter.

**Comment by emilgelman** [comment]
Thank you @KSerrania for the thorough explanation. We will raise an issue in the helper transformer.
